### PR TITLE
exploit php 7.1 features

### PR DESCRIPTION
**Pull request by garak**

### CODE REVIEW

File: src/Ulid.php [hunk 1]
```diff
@@ -41,39 +41,21 @@ class Ulid
...
-     * @param string $value
-     * @param bool $lowercase
-     *
-     * @return Ulid
-     */
-    public static function fromString($value, $lowercase = false)
+    public static function fromString(string $value, bool $lowercase = false): self
     {
-        return new static(substr($value, 0, 10), substr($value, 10), $lowercase);
+        return new static(\substr($value, 0, 10), \substr($value, 10), $lowercase);
```
Comment (robinvdvleuten) [inline src/Ulid.php:53]:
```suggestion
        return new static(substr($value, 0, 10), substr($value, 10), $lowercase);
```

File: src/Ulid.php [hunk 2]
```diff
@@ -41,39 +41,21 @@ class Ulid
...
-    /**
-     * @param bool $lowercase
-     *
-     * @return Ulid
-     */
-    public static function generate($lowercase = false)
+    public static function generate(bool $lowercase = false): self
     {
-        $now = intval(microtime(true) * 1000);
+        $now = \intval(\microtime(true) * 1000);
```
Comment (robinvdvleuten) [inline src/Ulid.php:58]:
```suggestion
        $now = intval(microtime(true) * 1000);
```

File: src/Ulid.php [hunk 3]
```diff
@@ -89,7 +71,7 @@ public static function generate($lowercase = false)
 
         if (!$duplicateTime) {
             for ($i = 0; $i < 16; $i++) {
-                static::$lastRandChars[$i] = random_int(0, 31);
+                static::$lastRandChars[$i] = \random_int(0, 31);
```
Comment (robinvdvleuten) [inline src/Ulid.php:74]:
```suggestion
                static::$lastRandChars[$i] = random_int(0, 31);
```

File: src/Ulid.php [hunk 4]
```diff
@@ -108,11 +90,8 @@ public static function generate($lowercase = false)
...
     }
 
-    /**
-     * @return string
-     */
-    public function __toString()
+    public function __toString(): string
     {
-        return ($value = $this->time . $this->randomness) && $this->lowercase ? strtolower($value) : strtoupper($value);
+        return ($value = $this->time . $this->randomness) && $this->lowercase ? \strtolower($value) : \strtoupper($value);
```
Comment (robinvdvleuten) [inline src/Ulid.php:95]:
```suggestion
        return ($value = $this->time . $this->randomness) && $this->lowercase ? strtolower($value) : strtoupper($value);
```

File: tests/UlidTest.php [hunk 5]
```diff
@@ -14,38 +14,38 @@
...
+        $this->assertSame(26, \strlen(Ulid::generate()));
     }
 
-    public function testGeneratesLexographicallySortableUlids()
+    public function testGeneratesLexographicallySortableUlids(): void
     {
         $a = Ulid::generate();
 
-        sleep(1);
+        \sleep(1);
```
Comment (robinvdvleuten) [inline tests/UlidTest.php:38]:
```suggestion
        sleep(1);
```

File: tests/UlidTest.php [hunk 6]
```diff
@@ -14,38 +14,38 @@
...
 
         $b = Ulid::generate();
 
         $ulids = [(string) $b, (string) $a];
-        usort($ulids, 'strcmp');
+        \usort($ulids, 'strcmp');
```
Comment (robinvdvleuten) [inline tests/UlidTest.php:43]:
```suggestion
        usort($ulids, 'strcmp');
```

File: tests/UlidTest.php [hunk 7]
```diff
@@ -14,38 +14,38 @@
...
+    public function testGeneratesLowercaseIdentiferWhenConfigured(): void
     {
         $this->assertRegExp('/[0-9][a-z]/', (string) Ulid::generate(true));
     }
 
-    public function testGeneratesTwentySixChars()
+    public function testGeneratesTwentySixChars(): void
     {
-        $this->assertSame(26, strlen(Ulid::generate()));
+        $this->assertSame(26, \strlen(Ulid::generate()));
```
Comment (robinvdvleuten) [inline tests/UlidTest.php:31]:
```suggestion
        $this->assertSame(26, strlen(Ulid::generate()));
```

**Review by robinvdvleuten — CHANGES_REQUESTED** [review]

**Comment by robinvdvleuten** [comment]
Thanks for the effort! But could you please remove all `\` from the global functions?

**Comment by garak** [comment]
@robinvdvleuten well, they are there on purpose: to avoid fallback on global namespace.
I know that it's just a micro-optimization, but why to remove it?

**Comment by robinvdvleuten** [comment]
They already are global functions and I can't imagine I would ever write the same named functions in the library's namespace.

**Comment by garak** [comment]
They are global functions used inside `Ulid`.
This means that PHP is looking, for example, for a function named `Ulid\randomint` and then, not founding its definition, make a fallback on `\randomint`.
Using the slash is meant to avoid such search & fallback

**Comment by robinvdvleuten** [comment]
Hmm ok, I checked it within some large PHP projects and apparently this is indeed a new thing to do in the PHP world.

**Pull request merged**